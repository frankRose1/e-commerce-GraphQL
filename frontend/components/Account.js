import React, { Component } from 'react';
import gql from 'graphql-tag';
import {Query} from 'react-apollo';
import {formatDistance} from 'date-fns';
import styled from 'styled-components';
import Link from 'next/link';
import formatMoney from '../lib/formatMoney';
import OrderItem from './styles/OrderItemStyles';
import Error from './ErrorMessage';
import Loading from './Loading';
import SubTitle from './styles/SubTitleStyles';

const OrdersGrid = styled.ul`
  display: grid;
  grid-gap: 4rem;
  grid-template-columns: repeat(auto-fit, minmax(40%, 1fr));
`;

const USER_ORDERS_QUERY = gql`
  query USER_ORDERS_QUERY {
    userOrders(orderBy: createdAt_DESC) {
      id
      total
      createdAt
      items {
        id
        quantity
        description
        title
        price
        image
      }
    }
  }
`;

class Account extends Component {
  render() {
    return (
      <Query query={USER_ORDERS_QUERY}>
        {({data, loading, error}) => {
          const {userOrders} = data;
          if (loading) return <Loading />
          if(error) return <Error error={error} />
          if (!userOrders.length) return <p>Looks like you havn't ordered anything yet!</p>
          
          return (
            <div>
              <SubTitle>Your Account</SubTitle>
              <h2>You have {userOrders.length} order{userOrders.length === 1 ? '' : 's'}</h2>
              <OrdersGrid>
                {userOrders.map(order => (
                  <OrderItem key={order.id}>
                    <Link href={{
                      pathname: '/order',
                      query: { orderId: order.id }
                    }}>
                      <a>
                        <div className="order-meta">
                          <p>{order.items.reduce((a, b) => a + b.quantity, 0)} Items</p>
                          <p>{order.items.length} Product{order.items.length === 1 ? '' : 's'}</p>
                          <p>{formatDistance(order.createdAt, new Date())}</p>
                          <p>{formatMoney(order.total)}</p>
                        </div>
                        <div className="images">
                          {order.items.map(item => (
                            <img 
                              src={item.image} 
                              alt={item.title}
                              key={item.id} />
                          ))}
                        </div>
                      </a>
                    </Link>
                  </OrderItem>
                ))}
            </OrdersGrid>
            </div>
          )
        }}
      </Query>
    );
  }
}

export default Account;