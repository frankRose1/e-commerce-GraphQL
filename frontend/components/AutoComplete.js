import React, { Component } from 'react';
import {ApolloConsumer} from 'react-apollo'; //run queries on demand
import gql from 'graphql-tag';
import Router from 'next/router';
import debounce from 'lodash.debounce';
import Downshift, {resetIdCounter} from 'downshift';
import {DropDown, DropDownItem, SearchStyles} from './styles/DropDown';

function routeToItem(item){
  Router.push({
    pathname: '/item',
    query: { id: item.id }
  })
}

//serach by either the title or description
const SEARCH_ITEMS_QUERY = gql`
  query SEARCH_ITEMS_QUERY($searchTerm: String!){
    items(where: {
      OR: [
        { title_contains: $searchTerm},
        { description_contains: $searchTerm}
      ]
    }){
      id
      image
      title
    }
  }
`;

class AutoComplete extends Component {

  state= {
    loading: false,
    items: []
  }

  //Manually fire off a query on input change
  handleChange = debounce(async (e, {query}) => {
    this.setState({loading: true});
    const res = await query({
      query: SEARCH_ITEMS_QUERY,
      variables: { searchTerm: e.target.value }
    });
    this.setState({
      loading: false,
      items: res.data.items
    })
  }, 400);

  render() {
    const {items, loading} = this.state;
    resetIdCounter();
    return (
      <SearchStyles>
        <Downshift
          onChange={routeToItem}
          itemToString={item => (item === null ? '' : item.title)}>
          { ({getInputProps, getItemProps, isOpen, inputValue, highlightedIndex}) => (
          <div>
            <ApolloConsumer>
              { client => (
                <input
                  {...getInputProps({
                    placeholder: "Search Our Awesome Collection",
                    type: "search",
                    id: 'search',
                    className: this.state.loading ? 'loading' : '',
                    onChange: e => {
                      e.persist();
                      this.handleChange(e, client);
                    }
                  })} />
              )}
            </ApolloConsumer>
            {isOpen && (
              <DropDown>
                {items.map((item, i) => (
                  <DropDownItem
                    {...getItemProps({item})}
                    highlighted={highlightedIndex === i}
                    key={item.id}>
                    <img width="50" src={item.image} alt={item.title}/>
                    {item.title}
                  </DropDownItem>
                ))}
                {!items.length && !loading && (
                  <DropDownItem>
                    Nothing found for "{inputValue}"
                  </DropDownItem>
                )}
              </DropDown>
            )}
          </div>
          )}
        </Downshift>
      </SearchStyles>
    );
  }
}

export default AutoComplete;