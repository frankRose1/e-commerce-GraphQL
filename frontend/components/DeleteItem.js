import React, { Component } from 'react';
import {Mutation} from 'react-apollo';
import gql from 'graphql-tag';
import PropTypes from 'prop-types';
import {ALL_ITEMS_QUERY} from './Items';

const DELETE_ITEM_MUTATION = gql`
  mutation DELETE_ITEM_MUTATION($id: ID!){
    deleteItem(id: $id){
      id
    }
  }
`;

class DeleteItem extends Component {

  static propTypes = {
    id: PropTypes.string.isRequired
  };

  //maually update the cache so that it matches the data on the server. and updating cache will update the UI
  updateCache = (cache, payload) => {
    //read the cache for the items. we need to do this wit the query used to fetch the items in the first place
    const data = cache.readQuery({ query: ALL_ITEMS_QUERY });
    // remove the deleted ID from cache
    data.items = data.items.filter(item => item.id !== payload.data.deleteItem.id);
    //put the items back
    cache.writeQuery({query: ALL_ITEMS_QUERY, data});
  }

  render() {
    return (
      <Mutation
        mutation={DELETE_ITEM_MUTATION}
        variables={{
          id: this.props.id
        }}
        update={this.updateCache}>
        {(deleteItem, {error, loading}) => (
          <button
          disabled={loading}
          onClick={() => {
            if (confirm('Are you sure you want to delete this item?')){
              deleteItem().catch(err => {
                alert(err.message);
              });
            }
          }}>
            Delete This Item
          </button>
        )}
      </Mutation>
    );
  }
}



export default DeleteItem;