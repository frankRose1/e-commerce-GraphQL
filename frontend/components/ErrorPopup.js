import React from 'react';
import PropTypes from 'prop-types';
import ErrPopUpStyles from './styles/ErrPopUpStyles';

//meant to handle errors on buttons where the UI is too small to display an error message

const ErrorPopup = ({error}) => {
  
  console.log(error);
  return (
    <ErrPopUpStyles showPopup={error ? true : false}>
      <p>{error}</p>
    </ErrPopUpStyles>
  );
};

ErrorPopup.propTypes = {
  error: PropTypes.string.isRequired,
};

export default ErrorPopup;