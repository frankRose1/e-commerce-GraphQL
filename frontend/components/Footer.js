import React from 'react';
import { FooterStyles, FooterLogo } from './styles/FooterStyles';
import Link from 'next/link';

const Footer = () => (
  <FooterStyles>
    <FooterLogo>BackPackzz</FooterLogo>
    <div className='footer-cols'>
      <div>
        <h4>Around The Web</h4>
        <ul>
          <li>
            <a href='https://www.facebook.com/' target='_blank'>
              {' '}
              <i className='fab fa-facebook-f' /> facebook
            </a>
          </li>
          <li>
            <a href='https://www.instagram.com/' target='_blank'>
              <i className='fab fa-instagram' /> instagram
            </a>
          </li>
          <li>
            <a href='https://www.twitter.com/' target='_blank'>
              <i className='fab fa-twitter' /> twitter
            </a>
          </li>
        </ul>
      </div>
      <div>
        <h4>Company</h4>
        <ul>
          <li>
            <Link href='/'>
              <a>Home</a>
            </Link>
          </li>
          <li>
            <Link href='/shop'>
              <a>Shop</a>
            </Link>
          </li>
          <li>
            <Link href='/'>
              <a>About Us</a>
            </Link>
          </li>
          <li>
            <Link href='/'>
              <a>Customer Service</a>
            </Link>
          </li>
        </ul>
      </div>
      <div>
        <h4>NewsLetter</h4>
        <p>
          Subscribe to get special offers, free giveaways, and
          once-in-a-lifetime deals!(psst! this is just a placeholder!)
        </p>
        <div className='subscribe'>
          <input type='email' placeholder='your-email@example.com' />
          <input type='submit' value='SUBSCRIBE' />
        </div>
      </div>
    </div>
    <p>&copy; BackPackzz {new Date().getFullYear()} | All rights Reserved</p>
  </FooterStyles>
);

export default Footer;
