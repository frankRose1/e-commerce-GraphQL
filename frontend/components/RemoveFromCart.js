import React, { Component } from 'react';
import styled from 'styled-components';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import PropTypes from 'prop-types';
import { CURRENT_USER_QUERY } from './User';

//TODO: remove alerts for a popup here, and on delete item button

const BigButton = styled.button`
  font-size: 5rem;
  padding: 1.5rem;
  background: none;
  border: 0;
  outline: none;
  transition: color 0.4s ease-in-out;
  cursor: pointer;
  &:hover{
    color: ${props => props.theme.red}
  }
`;

const REMOVE_FROM_CART_MUTATION = gql`
  mutation REMOVE_CART_ITEM_MUTATION($cartItemId: ID!){
    removeFromCart(cartItemId: $cartItemId){
      id
    }
  }
`;

class RemoveFromCart extends Component {

  //When an item is removed, delete it from the cache as it takes a second to display in the UI when fetching to the DB
  //this is called after a response is recieved from the server after running a mutation
  //payload is the data requested in the mutation (only asked for cart item ID)
  updateCache = (cache, payload) => {
    const data = cache.readQuery({
      query: CURRENT_USER_QUERY
    });
    const id = payload.data.removeFromCart.id;
    data.me.cart = data.me.cart.filter(item => item.id !== id);
    cache.writeQuery({query: CURRENT_USER_QUERY, data});
  }

  render() {
    const {cartItemId} = this.props;
    //the optmistic repsonse will run immediately so that the UI can be updated quicker, while in the background the item is still being removed in the DB
    return (
      <Mutation 
        mutation={REMOVE_FROM_CART_MUTATION}
        variables={{ cartItemId }}
        update={this.updateCache}
        optimisticResponse={{
          __typename: 'Mutation',
          removeFromCart: {
            __typename: 'CartItem',
            id: cartItemId
          }
        }}>
        {(removeFromCart, {error, loading}) => (
          <BigButton
            disabled={loading}
            title="Remove Item"
            onClick={() => {
              removeFromCart().catch(err => alert(err.message));
            }}>
              &times;
          </BigButton>
        )}
      </Mutation>
    );
  }
}

RemoveFromCart.propTypes = {
  cartItemId: PropTypes.string.isRequired
}

export default RemoveFromCart;