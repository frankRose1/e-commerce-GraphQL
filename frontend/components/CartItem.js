import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import formatMoney from '../lib/formatMoney';
import RemoveFromCart from './RemoveFromCart';

const CartItemStyles = styled.li`
  padding: 1rem 0;
  display: grid;
  border-bottom: 1px solid ${props => props.theme.lightgrey};
  grid-template-columns: auto 1fr auto;
  align-items: center;
  img {
    width: 100px;
    margin-right: 10px;
  }
  h3, p {
    margin: 0;
  }
`;

const DeletedItem = styled.li`
  padding: 1rem 0;
  display: grid;
  border-bottom: 1px solid ${props => props.theme.lightgrey};
  grid-template-columns: 1fr auto;
  align-items: center;
`;

const CartItem = ({cartItem: { item, quantity, id} }) => {
  //in the case that An item is deleted in the DB we dont want to show it in the users cart
  if (!item){
    return (
      <DeletedItem>
        <p>Sorry, this item is no longer sold in our store!</p>
        <RemoveFromCart cartItemId={id}/>
      </DeletedItem>
    )
  }

  return (
  <CartItemStyles>
    <img src={item.image} alt={item.title}/>
    <div className="cart-item-details">
      <h3>{item.title}</h3>
      <p>
        {formatMoney(item.price * quantity)}
        {' - '}
        <em>{quantity} &times; {formatMoney(item.price)} each</em>
      </p>
    </div>
    <RemoveFromCart cartItemId={id}/>
  </CartItemStyles>
)};

CartItem.propTypes = {
  cartItem: PropTypes.shape({
    quantity: PropTypes.number.isRequired,
    id: PropTypes.string.isRequired,
    item: PropTypes.object
  }).isRequired
};

export default CartItem;