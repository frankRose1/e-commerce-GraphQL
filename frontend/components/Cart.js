import React from 'react';
import {Mutation, Query} from 'react-apollo';
import gql from 'graphql-tag';
import {adopt} from 'react-adopt';
import User from './User';
import CartItem from './CartItem';
import CartStyles from './styles/CartStyles';
import Supreme from './styles/Supreme';
import CloseButton from './styles/CloseButton';
import Button from './styles/SickButton';
import PurchaseItem from './PurchaseItem';
import calcTotalPrice from '../lib/calcTotalPrice';
import formatMoney from '../lib/formatMoney';

//Client side query & mutation for cart boolean
const LOCAL_STATE_QUERY = gql`
  query {
    cartOpen @client
  }
`;

const TOGGLE_CART_MUTATION = gql`
  mutation {
    toggleCart @client
  }
`;

const Composed = adopt({
  user: ({render}) => <User>{render}</User>,
  toggleCart: ({render}) => <Mutation mutation={TOGGLE_CART_MUTATION}>{render}</Mutation>,
  localState: ({render}) => <Query query={LOCAL_STATE_QUERY}>{render}</Query>
});

const Cart = () => {
  return (
    <Composed>
      {( {user, toggleCart, localState} ) => {
        const {data: {me}, loading} = user;
        if (loading) return <p>Loading...</p>
        if(!me) return null //dont show the cart if user is not signed in
        return (
          <CartStyles
            data-test="cart"
            open={localState.data.cartOpen}>
            <header>
              <CloseButton onClick={toggleCart} title="close">&times;</CloseButton>
              <Supreme>{me.name}'s Cart</Supreme>
              <p>{me.cart.length} item{me.cart.length === 1 ? '' : 's'} in your cart</p>
            </header>

              <ul>
                {me.cart.map(cartItem => (
                  <CartItem 
                    key={cartItem.id}
                    cartItem={cartItem}/>
                ))}
              </ul>

            <footer>
              <p>{formatMoney(calcTotalPrice(me.cart))}</p>
              {me.cart.length
                ? (
                  <PurchaseItem>
                    <Button>Checkout</Button>
                  </PurchaseItem>
                )
              : null }
            </footer>
          </CartStyles>
        )
      }}
    </Composed>
  );
};

export default Cart;
export {TOGGLE_CART_MUTATION, LOCAL_STATE_QUERY};