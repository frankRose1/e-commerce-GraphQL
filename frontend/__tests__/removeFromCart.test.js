import {mount} from 'enzyme';
import {MockedProvider} from 'react-apollo/test-utils';
import toJSON from 'enzyme-to-json';
import {ApolloConsumer} from 'react-apollo';
import wait from 'waait';
import RemoveFromCart, {REMOVE_FROM_CART_MUTATION} from '../components/RemoveFromCart';
import {CURRENT_USER_QUERY} from '../components/User';
import {fakeUser, fakeCartItem} from '../lib/testUtils';

/**
 * RemoveFromCart has an update function that will run once the mutation sends back data
 * dont need to write a second user query mock because it is handled manually in the update
 */

const mocks = [
  //current user for when the component mounts
  {
    request: {query: CURRENT_USER_QUERY },
    result: {
      data: { me: {
        ...fakeUser(),
        cart: [fakeCartItem({id: 'abc123'})]
      }}
    },
  },
  {
    request: {query: REMOVE_FROM_CART_MUTATION, variables: { cartItemId: 'abc123'}},
    result: {
      data: {
        removeFromCart : {
          __typename: 'CartItem',
          id: 'abc123'
        }
      }
    },
  },
];

describe('<RemoveFromCart />', () => {

  it("renders and matches snapshot", () => {
    const wrapper = mount(
      <MockedProvider>
        <RemoveFromCart cartItemId="abc123" />
      </MockedProvider>
    );
    expect(toJSON(wrapper.find('button'))).toMatchSnapshot();
  });

  it("removes item from cart after button is clicked", async () => {
    let apolloClient;
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <ApolloConsumer>
          {client => {
            apolloClient = client;
            return  <RemoveFromCart cartItemId="abc123" />
          }}
        </ApolloConsumer>
      </MockedProvider>
    );
    const {data : { me } } = await apolloClient.query({query: CURRENT_USER_QUERY });
    expect(me.cart).toHaveLength(1);
    expect(me.cart[0].item.price).toBe(5000);
    //simulate clicking the button and removing the cart Item
    wrapper.find('button').simulate('click');
    await wait();
    const {data: { me: me2 } } = await apolloClient.query({query: CURRENT_USER_QUERY });
    expect(me2.cart).toHaveLength(0);
  });
});