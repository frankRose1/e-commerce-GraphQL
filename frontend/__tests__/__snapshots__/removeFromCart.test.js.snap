// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RemoveFromCart /> renders and matches snapshot 1`] = `
<button
  className="RemoveFromCart__BigButton-r0391h-0 jmuTzT"
  disabled={false}
  onClick={[Function]}
  title="Remove Item"
>
  Ã—
</button>
`;
